autoprecompiles/src/equivalence_classes.rs

```diff
@@ -2,74 +2,167 @@ use std::collections::{BTreeSet, HashMap};
+impl<T: Eq + Hash + Serialize + Clone> Serialize for Partition<T> {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        // Serialize as Vec<Vec<T>> for JSON compatibility
+        self.to_classes().serialize(serializer)
+    }
+}
+
+impl<'de, T: Eq + Hash + Deserialize<'de>> Deserialize<'de> for Partition<T> {
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        // Deserialize from Vec<Vec<T>>, reusing FromIterator logic
+        let classes: Vec<Vec<T>> = Vec::deserialize(deserializer)?;
+        Ok(classes.into_iter().collect())
+    }
+}
+
+impl<T: Eq + Hash, C: IntoIterator<Item = T>> FromIterator<C> for Partition<T>
+where
+    C::IntoIter: ExactSizeIterator,
+{
+    fn from_iter<I: IntoIterator<Item = C>>(iter: I) -> Self {
+        let mut class_of = HashMap::new();
+        let mut num_classes = 0;
+
+        for class in iter {
+            let class_iter = class.into_iter();
+            // Ignore classes with 0 or 1 elements as they are useless
+            if class_iter.len() > 1 {
+                for element in class_iter {
+                    assert!(class_of.insert(element, num_classes).is_none());
+                }
+                num_classes += 1;
+            }
+        }
+
         Self {
-            inner: iter.into_iter().filter(|class| class.len() > 1).collect(),
+            class_of,
+            num_classes,
         }
     }
 }
 
-impl<T> Partition<T> {
-    pub fn iter(&self) -> impl Iterator<Item = &EquivalenceClass<T>> {
-        self.inner.iter()
+impl<T: Eq + Hash + Clone> Partition<T> {
+    /// Returns all equivalence classes as a Vec<Vec<T>>.
+    /// This is O(n) where n is the number of elements.
+    #[allow(clippy::iter_over_hash_type)] // Order within classes doesn't matter semantically
+    pub fn to_classes(&self) -> Vec<Vec<T>> {
+        let mut classes: Vec<Vec<T>> = vec![Vec::new(); self.num_classes];
+        for (elem, &class_id) in &self.class_of {
+            classes[class_id].push(elem.clone());
+        }
+        classes
+    }
+
+    pub fn empty() -> Self {
+        Self {
+            class_of: HashMap::new(),
+            num_classes: 0,
+        }
     }
-}
 
-impl<T: Ord + Hash + Copy> Partition<T> {
     /// Intersects multiple partitions of the same universe into a single partition.
     /// In other words, two elements are in the same equivalence class in the resulting partition
     /// if and only if they are in the same equivalence class in all input partitions.
     /// Singleton equivalence classes are omitted from the result.
-    pub fn intersect(partitions: &[Self]) -> Self {
-        // For each partition, build a map: Id -> class_index
-        let class_ids: Vec<HashMap<T, usize>> = partitions
+    pub fn intersect_many(partitions: &[Self]) -> Self {
+        let Some(first) = partitions.first() else {
```

Comment:
Can we simplify this by folding the entire slice starting with `Self::empty()`? Or does intersection break with empty partitions?

Reply:
Yes, an empty partition would mean each element is a singleton. Actually, maybe I'll just panic if `intersection_many` is called with an empty slice, it should never happen...